File: library-appcompat/src/main/java/com/cz/android/sample/library/permission/PermissionViewModelProviders.java
package com.cz.android.sample.library.permission;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.ViewModelProvider;

/**
 * @author Created by cz
 * @date 2020/7/28 5:02 PM
 * @email bingo110@126.com
 */
public class PermissionViewModelProviders{

    public static PermissionViewModel getViewModel(FragmentActivity activity){
        ViewModelProvider.Factory newInstanceFactory = new ViewModelProvider.NewInstanceFactory();
        ViewModelProvider viewModelProvider = new ViewModelProvider(activity.getViewModelStore(), newInstanceFactory);
        return viewModelProvider.get(PermissionViewModel.class);
    }

    /**
     * We will only use the activity's viewModelStore to store the ViewModel.
     * If we use the different lifecycleOwner with different ViewModelStore.
     * The fragment will never receive the event.
     * @param fragment
     * @return
     */
    public static PermissionViewModel getViewModel(Fragment fragment){
        FragmentActivity activity = fragment.getActivity();
        ViewModelProvider.Factory newInstanceFactory = new ViewModelProvider.NewInstanceFactory();
        ViewModelProvider viewModelProvider = new ViewModelProvider(activity.getViewModelStore(), newInstanceFactory);
        return viewModelProvider.get(PermissionViewModel.class);
    }

}

File: extension/src/main/java/com/github/jackchen/android/sample/library/function/permission/PermissionViewModelProviders.kt
package com.github.jackchen.android.sample.library.function.permission

import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.Observer
import com.github.jackchen.android.sample.library.function.permission.PermissionViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProvider.NewInstanceFactory
import androidx.lifecycle.ViewModelStoreOwner


fun FragmentActivity.addPermissionObserver(observer: Observer<PermissionResult>) {
    PermissionViewModelProviders.getViewModel(this).addObserver(this, observer)
}

fun Fragment.addPermissionObserver(observer: Observer<PermissionResult>) {
    PermissionViewModelProviders.getViewModel(this).addObserver(this, observer)
}

object PermissionViewModelProviders {
    fun getViewModel(activity: FragmentActivity): PermissionViewModel {
        val newInstanceFactory: ViewModelProvider.Factory = NewInstanceFactory()
        val viewModelProvider = ViewModelProvider(activity, newInstanceFactory)
        return viewModelProvider.get(PermissionViewModel::class.java)
    }

    /**
     * We will only use the activity's viewModelStore to store the ViewModel.
     * If we use the different lifecycleOwner with different ViewModelStore.
     * The fragment will never receive the event.
     *
     * @param fragment
     * @return
     */
    fun getViewModel(fragment: Fragment): PermissionViewModel {
        val activity = fragment.activity
        val newInstanceFactory: ViewModelProvider.Factory = NewInstanceFactory()
        val viewModelProvider = ViewModelProvider(activity!!.viewModelStore, newInstanceFactory)
        return viewModelProvider.get(PermissionViewModel::class.java)
    }
}